### PR TITLE
Adds some convenient smt formula construction classes

### PR DESCRIPTION
**Pull request by rdaly525**

### CODE REVIEW

File: hwtypes/smt_utils.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+def _op_to_str(vs, opname, ts, indent):
+    new_ts = ts + indent
+    return "\n".join([
+        f"{ts}{opname}(",
+        ",\n".join([_value_to_str(v, new_ts, indent) for v in vs]),
+        f"{ts})"
+    ])
+def _check(vs):
+    if not all(isinstance(v, FormulaConstructor) or isinstance(v, SMTBit) for v in vs):
```
Comment (cdonovick) [inline hwtypes/smt_utils.py]: `isinstance(v, (FormulaConstructor, SMTBit))`
Performs:
`isinstance(v, FormulaConstructor) or isinstance(v, SMTBit))`

For reference `issubclass` also can take a tuple of types.

File: tests/test_smt_utils.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    f_ht = f.to_hwtypes()
+    assert isinstance(f_ht, SMTBit)
+    f_str = f.serialize()
+    assert isinstance(f_str, str)
+
+
+def test_0_len():
+    f = utils.And([])
+    assert f.to_hwtypes().value.constant_value() == True
```
Comment (cdonovick) [inline tests/test_smt_utils.py]: `== True` is unnecessary unless you are trying to check that it actually returns a bool in which case I would use `is` over `==`.

File: tests/test_smt_utils.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    f = utils.Or([])
+    assert f.to_hwtypes().value.constant_value() == False
```
Comment (cdonovick) [inline tests/test_smt_utils.py]: Use `not` over `== False` or similar to above `is`.

File: tests/test_smt_utils.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        ]),
+        SMTBit(1)
+    )
+
+    f_ht = f.to_hwtypes()
+    assert isinstance(f_ht, SMTBit)
+    f_str = f.serialize()
+    assert isinstance(f_str, str)
+
+
```
Comment (cdonovick) [inline tests/test_smt_utils.py:48]: Add a test for serialize and its arguments against expected output.

File: hwtypes/smt_utils.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+import operator
+from functools import partial, reduce
+import abc
+
+_or_reduce = partial(reduce, operator.or_)
+_and_reduce = partial(reduce, operator.and_)
+
+class FormulaConstructor:
+    @abc.abstractmethod
+    def serialize(self, ts, indent):
```
Comment (cdonovick) [inline hwtypes/smt_utils.py]: I would prefer a visitor pattern with serialization visitor but I can live with this,

File: hwtypes/smt_utils.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            return SMTBit(False)
+        return _or_reduce(_to_hwtypes(v) for v in self.values)
+
+class Implies(FormulaConstructor):
+    def __init__(self, p, q):
+        _check([p, q])
+        self.p = p
+        self.q = q
+
+    def serialize(self, ts="", indent="|   "):
```
Comment (cdonovick) [inline hwtypes/smt_utils.py]: What is `ts`

**Review by cdonovick — CHANGES_REQUESTED** [review]

File: hwtypes/smt_utils.py (hunk 6 above)
Comment (rdaly525) [reply hwtypes/smt_utils.py]: "Tab space". Basically the accumulation of indents. I wrote the pretty serialization code a while ago and it works well. I prefer not to touch it. I will add a test though.

File: tests/test_smt_utils.py (hunk 3 above)
Comment (rdaly525) [reply tests/test_smt_utils.py]: Fixed

File: tests/test_smt_utils.py (hunk 2 above)
Comment (rdaly525) [reply tests/test_smt_utils.py]: Fixed

File: hwtypes/smt_utils.py (hunk 1 above)
Comment (rdaly525) [reply hwtypes/smt_utils.py]: Fixed

File: hwtypes/smt_utils.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+def _op_to_str(vs, opname, ts, indent):
+    new_ts = ts + indent
+    return "\n".join([
+        f"{ts}{opname}(",
+        ",\n".join([_value_to_str(v, new_ts, indent) for v in vs]),
+        f"{ts})"
+    ])
+def _check(vs):
+    if not all(isinstance(v, (FormulaConstructor, SMTBit)) for v in vs):
```
Comment (cdonovick) [inline hwtypes/smt_utils.py:37]: doesn't really matter but `any(not ...)` is more efficient than `not all(...)`

File: hwtypes/smt_utils.py (hunk 6 above)
Comment (cdonovick) [reply hwtypes/smt_utils.py]: So its not really part of the api but rather for recursion? 

I would very much prefer a slight refactor:
```
def serialize(self, indent="| "): 
    self._serialize(indent)
    
def _serialize(self, indent, ts=''): 
    #current code here. 
```

Could even put  the above implementation of `serialize` in the ABC and make `_serialize` abstract.  

If `ts` is meant to be part of the API please add a test.

File: hwtypes/smt_utils.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
+from . import SMTBit
+import operator
+from functools import partial, reduce
+import abc
+
+_or_reduce = partial(reduce, operator.or_)
+_and_reduce = partial(reduce, operator.and_)
+
+class FormulaConstructor:
```
Comment (cdonovick) [inline hwtypes/smt_utils.py]: For abc.abstractmethod to do anything you need to inherit from `abc.ABC` or use the abc metaclass `metaclass=abc.ABCMeta` (which is what inheriting from abc.ABC does).

File: tests/test_smt_utils.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+                ~z & x,
+            ]),
+        ]),
+        SMTBit(1)
+    )
+
+    f_ht = f.to_hwtypes()
+    assert isinstance(f_ht, SMTBit)
+    f_str = f.serialize()
+    print(f_str)
```
Comment (cdonovick) [inline tests/test_smt_utils.py]: remove print

File: hwtypes/smt_utils.py (hunk 6 above)
Comment (cdonovick) [reply hwtypes/smt_utils.py]: Oops seems like you pushed code while I was reviewing so not sure if above is still fully important.  But still like the refactor if you don't actually use the `ts` arg nor expect people to.

Comment (rdaly525) [reply hwtypes/smt_utils.py]: ts can be used as the initial space offset (as seen in the recent test). Could be useful.

Comment (cdonovick) [reply hwtypes/smt_utils.py]: I will be pressing x to doubt on that one.  If its part of the api it will need a more descriptive name (e.g. `prefix`).

**Review by cdonovick — APPROVED** [review]

**Pull request merged**